Sming/system/m_printf.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+Author: (github.com/)ADiea
+Project: Sming for ESP8266 - https://github.com/anakod/Sming
+License: MIT
+Date: 20.07.2015
+Descr: embedded very simple version of printf with float support
+*/
+
+#include <stdarg.h>
+#include "../../SmingCore/SmingCore.h"
+
+#define MPRINTF_BUF_SIZE 256
+
+void (*cbc_printchar)(char ch) = uart_tx_one_char;
+
+#define SIGN    	(1<<1)	/* Unsigned/signed long */
+
+int m_vsnprintf(char *buf, uint32_t maxLen, const char *fmt, va_list args);
+
+void setMPrintfPrinterCbc(void (*callback)(char))
+{
+	cbc_printchar = callback;
+}
+
+int m_printf(const char *fmt, ...)
+{
+	char buf[MPRINTF_BUF_SIZE], *p;
+	va_list args;
+	int n = 0;
+
+	va_start(args, fmt);
+	m_vsnprintf(buf, sizeof(buf), fmt, args);
+	va_end(args);
+
+	p = buf;
+	while (*p)
+	{
+		cbc_printchar(*p);
+		n++;
+		p++;
+	}
+
+	return n;
+}
+
+int m_vsnprintf(char *buf, uint32_t maxLen, const char *fmt, va_list args)
+{
+	int i, base, flags, qualifier;
+	char *str;
+	const char *s;
+	uint8_t precision = 6;
+
+	const uint8_t overflowGuard = 24;
+
+	char tempNum[24];
+
+	for (str = buf; *fmt; fmt++)
+	{
+		if(maxLen - (str - buf) < overflowGuard)
+		{
+			*str++ = '.';
+			*str++ = '.';
+			*str++ = '.';
+			break;
+		}
+
+		if (*fmt != '%')
+		{
+			*str++ = *fmt;
+			continue;
+		}
+
+		flags = 0;
+		fmt++; // This skips first '%'
+
+		do
+		{
+			//skip width and flags data - not supported
```

Comment:
I believe they are all pretty print fluffs...if you feel we should support something let me know.

for ints: 
* no width support -> needed probably when displaying table data, i think rarely
* no left adjustment support -> rarely used, maybe for tables
* hex numbers always display 0x prefix to avoid confusion (printf has the # flag)
* the star flag saying that width and precision info are taken from parameters and not from format string -> rarely used
* no zero padding in front of int -> maybe usefull for table and pretty printing?

for floats
* presision is supprted as it is useful. 
* width not supported, only useful probably for pretty printing
* zero padding / space padding related to width above not supported

I could add width specifing for floats, unfortunately the ltoa function used here for ints does not take width parameter and it is already used like this in the project